.circleci/config.yml

```diff
@@ -48,10 +48,23 @@ version: 2.1
   attach_workspace:
     at: spec/rails
 
+.restore_test_times: &restore_test_times
+  restore_cache:
+    keys:
+      - testtime-{{ .Branch }}-{{ .Environment.CIRCLE_JOB }}
+      - testtime-master-{{ .Environment.CIRCLE_JOB }}
```

Comment:
What's the reason for restoring both dynamically from a feature branch and master? Not that this is wrong, but just by looking at how we save times, I didn't understand.

Reply:
What parallel_tests does is split all the tests in groups, and run each group in parallel. If you don't specify any criteria, it will split in groups of the same number of tests randomly. That means, for example, that all integration specs could fall in one group, and all model specs in another. And that means the group with the integration tests will be a bottleneck.

To solve this issue, and ideally have no bottlenecks, parallel_tests can optionally record the time it took for each test to run, and then in subsequent runs reuse that information to find the optimal split between tests so that all groups take approximately the same time. So this is what we're doing here, record that information so that our builds always split tests into groups optimally.

The edge case is that on the first push to a branch we will have no recorded information for that branch. If that is the case, we look at what was last recorded for the master branch, which in most cases will still be a good approximation. This `keys:` section does not mean we're restoring both, it means we're first looking at the first key, and it that's not present in the cache, we look for the second one.

Hope that all makes sense!

Reply:
Yes that does! It makes a lot of sense. I realize now I've done basically this but on what was CircleCI 1.0 as they had a tool for splitting tests which essentially seemed to just pick sets of spec files and specified that list to individual Rspec commands that would run. I remember about saving the timings of that too. The timings here come from the format option: `--format ParallelTests::RSpec::RuntimeLogger --out tmp/parallel_runtime_rspec.log` which is what I had to use as well but with RspecJunitFormatter and the `circleci tests` command on their older platform. Thanks for clarifying! It helped as I now know it's all doing essentially the same thing. 👍🏻